File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/FirstAggregationResult.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.FloatResultNode;
import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.searchlib.expression.SingleResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an aggregated result holding the value of the aggregating expression for the first hit in the group.
 *
 * The first hit is the one with the highest hit rank; ties keep the hit that was aggregated first. The rank of that
 * hit is part of the result, so that merging partial results from several nodes picks the same hit no matter what
 * order they merge in. Ties between equal ranks are resolved by merge order, and are unspecified.
 *
 * @author johsol
 */
public class FirstAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 183, FirstAggregationResult.class, FirstAggregationResult::new);
    private SingleResultNode first = new FloatResultNode(0.0);
    private double hitRank = Double.NEGATIVE_INFINITY;
    private boolean hasValue = false;

    /**
     * Constructs an empty result node.
     */
    public FirstAggregationResult() {}

    /**
     * Constructs an instance of this class with the given value, and no hit rank.
     *
     * @param first The value of the first hit.
     */
    public FirstAggregationResult(SingleResultNode first) {
        setFirst(first);
    }

    /**
     * Constructs an instance of this class with the given value and hit rank.
     *
     * @param first The value of the first hit.
     * @param hitRank The rank of the first hit.
     */
    public FirstAggregationResult(SingleResultNode first, double hitRank) {
        setFirst(first);
        setHitRank(hitRank);
    }

    /**
     * Returns the value of the first hit in this.
     *
     * @return The value of the first hit.
     */
    public final SingleResultNode getFirst() {
        return first;
    }

    /**
     * Sets the value of the first hit in this, marking this as holding a value.
     *
     * @param first The value to set.
     * @return This, to allow chaining.
     */
    public final FirstAggregationResult setFirst(SingleResultNode first) {
        this.first = first;
        this.hasValue = true;
        return this;
    }

    /**
     * Returns the rank of the hit the value was taken from.
     *
     * @return The hit rank.
     */
    public final double getHitRank() {
        return hitRank;
    }

    /**
     * Sets the rank of the hit the value was taken from.
     *
     * @param hitRank The hit rank to set.
     * @return This, to allow chaining.
     */
    public final FirstAggregationResult setHitRank(double hitRank) {
        this.hitRank = hitRank;
        return this;
    }

    /**
     * Returns whether any hit has been aggregated into this.
     *
     * @return True if this holds the value of a hit.
     */
    public final boolean hasValue() {
        return hasValue;
    }

    @Override
    public ResultNode getRank() {
        return first;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        buf.putByte(null, hasValue ? (byte)1 : (byte)0);
        buf.putDouble(null, hitRank);
        serializeOptional(buf, first);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        hasValue = buf.getByte(null) != 0;
        hitRank = buf.getDouble(null);
        first = (SingleResultNode)deserializeOptional(buf);
    }

    @Override
    protected void onMerge(AggregationResult result) {
        FirstAggregationResult rhs = (FirstAggregationResult)result;
        if (!rhs.hasValue || (hasValue && rhs.hitRank <= hitRank)) {
            return;
        }
        first = (SingleResultNode)rhs.first.clone();
        hitRank = rhs.hitRank;
        hasValue = true;
    }

    @Override
    public FirstAggregationResult clone() {
        FirstAggregationResult obj = (FirstAggregationResult)super.clone();
        if (first != null) {
            obj.first = (SingleResultNode)first.clone();
        }
        return obj;
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        FirstAggregationResult rhs = (FirstAggregationResult)obj;
        return hasValue == rhs.hasValue && hitRank == rhs.hitRank && equals(first, rhs.first);
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("hasValue", hasValue);
        visitor.visit("hitRank", hitRank);
        visitor.visit("first", first);
    }
}

File: searchlib/src/vespa/searchlib/aggregation/first_aggregation_result.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "first_aggregation_result.h"

#include <vespa/searchlib/expression/floatresultnode.h>
#include <vespa/searchlib/expression/resultvector.h>

#include <vespa/vespalib/objects/visit.hpp>

namespace search::aggregation {

using expression::FloatResultNode;
using expression::ResultNodeVector;
using expression::SingleResultNode;

using vespalib::Deserializer;
using vespalib::Serializer;

IMPLEMENT_IDENTIFIABLE_NS2(search, aggregation, FirstAggregationResult, AggregationResult);

namespace {

[[nodiscard]] bool is_ready(const expression::ResultNode* current, const expression::ResultNode& wanted) noexcept {
    return current != nullptr && current->getClass().id() == wanted.getClass().id();
}

[[nodiscard]] std::unique_ptr<SingleResultNode> create_result_node(const expression::ResultNode& result) {
    auto base = result.createBaseType();
    if (dynamic_cast<SingleResultNode*>(base.get()) != nullptr) {
        return std::unique_ptr<SingleResultNode>(static_cast<SingleResultNode*>(base.release()));
    }
    return std::make_unique<FloatResultNode>();
}

} // namespace

FirstAggregationResult::FirstAggregationResult()
    : AggregationResult(), _first(FloatResultNode(0.0)), _hit_rank(default_rank_value), _has_value(false) {
}

FirstAggregationResult::FirstAggregationResult(const SingleResultNode& first, HitRank hit_rank)
    : AggregationResult(), _first(first), _hit_rank(hit_rank), _has_value(true) {
}

FirstAggregationResult::~FirstAggregationResult() = default;

void FirstAggregationResult::onPrepare(const ResultNode& result) {
    if (!is_ready(_first.get(), result)) {
        _first = create_result_node(result);
    }
}

void FirstAggregationResult::initForUnitTest(const ResultNode& result) {
    onPrepare(result);
    _first->set(result);
    _has_value = true;
}

void FirstAggregationResult::update_first(const ResultNode& result, HitRank rank) {
    if (_has_value && rank <= _hit_rank) {
        return;
    }
    if (result.isMultiValue()) {
        const auto& values = static_cast<const ResultNodeVector&>(result);
        if (values.empty()) {
            return;
        }
        _first->set(values.get(0));
    } else {
        _first->set(result);
    }
    _hit_rank = rank;
    _has_value = true;
}

void FirstAggregationResult::onAggregate(const ResultNode& result) {
    update_first(result, zero_rank_value);
}

void FirstAggregationResult::onAggregate(const ResultNode& result, DocId, HitRank rank) {
    update_first(result, rank);
}

void FirstAggregationResult::onAggregate(const ResultNode& result, const document::Document&, HitRank rank) {
    update_first(result, rank);
}

void FirstAggregationResult::onMerge(const AggregationResult& b) {
    const auto& rhs = static_cast<const FirstAggregationResult&>(b);
    if (!rhs._has_value || (_has_value && rhs._hit_rank <= _hit_rank)) {
        return;
    }
    _first->set(*rhs._first);
    _hit_rank = rhs._hit_rank;
    _has_value = true;
}

void FirstAggregationResult::onReset() {
    _first.reset(static_cast<SingleResultNode*>(_first->getClass().create()));
    _hit_rank = default_rank_value;
    _has_value = false;
}

Serializer& FirstAggregationResult::onSerialize(Serializer& os) const {
    AggregationResult::onSerialize(os);
    return os << _has_value << _hit_rank << _first;
}

Deserializer& FirstAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    return is >> _has_value >> _hit_rank >> _first;
}

void FirstAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "hasValue", _has_value);
    visit(visitor, "hitRank", _hit_rank);
    visit(visitor, "first", _first);
}

} // namespace search::aggregation
